keps/sig-api-machinery/1929-built-in-default/README.md

```diff
@@ -0,0 +1,480 @@
+changing the current mechanisms:
+
+- We don't use OpenAPI to drive validation nor defaulting
+currently. There is no routing of the OpenAPI to most operations in
+kubernetes (the only thing right now is server-side apply).
+- OpenAPI fields don't strictly map to Go types fields
+(e.g. capitalization, embedding)
+- Reflection can't access comments, only structTags (which is why we have
+both structTags AND comments for SMP, but that's ugly and suboptimal).
+
+One possible way to address that second point is to serialize the object
+back to json, apply the default, and re-deserialize back into a
+go-struct, but that's memory intensive and Kubernetes doesn't need that.
+
+We're proposing to use the same mechanisms that we've put in place for
+server-side apply, implemented in
+"sigs.k8s.io/structured-merge-diff". We currently have a way for
+server-side apply to convert Go types into "json"-like values without
+serializing/deserializing and allocating memory by using reflection. We
+also have mechanisms in place to parse, process and use the OpenAPI.
+
+In other words, the plan would be the following:
+1. Update kube-openapi to parse the `+default` marker and include it in
+the OpenAPI definition
+2. Add additional "default" field to sigs.k8s.io/structured-merge-diff
+internal schema, and parse that field from the OpenAPI.
+3. Create a new "AddDefault" method in sigs.k8s.io/structured-merge-diff
+that receives a Value (value is an interface that can be backed by a
+reflection based Go-type, associated with the corresponding
+OpenAPI). That method would add the defaults in-place.
+4. Add and wire things so that this function can be called prior to
+running the existing defaulting code.
+
+### Marker format
+
+The marker will have the following format:
+```
+// +default=<any>
+```
+which describes the value that the server will use if the field has a
+zero-value.
+
+`<any>` has to be in one-line YAML format (including its json form),
+representing the actual value to use by default. That is different from
+the current encoding in kubebuilder, but there seems to be consensus
+that they would be able to change that to YAML too for that new marker.
+
+Namely, this will allow to preserve the concise semantics that you want
+for most values:
+
+```golang
+type Foo struct {
+  // +default=32
+  Integer int
+
+  // +default=bar
+  String string
+
+  // +default=[popcorn, chips]
+  StringList []string
+}
+```
+while still allowing more complex expressions:
+```
+// +default=[{port: 80, name: http}]
+```
+
+Forcing the "one-line" subset allows us to sidestep a lot of the YAML
+weirdness in syntax, but we can still just use a YAML parser to parse
+the expressions if we want.
+
+### CRD Normalization
+
+In order to have a consistent behavior between CRDs and built-in, the
+following two things need to happen:
+
+1. Using the flag with CRDs requires the presence of a
+`x-kubernetes-remove-nulls` flag set to true at the root level of the
```

Comment:
and false is forbidden? Why only on root levels? Note that we can have types like PodSpec embedded into some CRD.